src/librustc_mir/transform/const_prop.rs

```diff
@@ -497,6 +497,67 @@ impl<'a, 'mir, 'tcx> ConstPropagator<'a, 'mir, 'tcx> {
             },
         }
     }
+
+    fn replace_with_const(&self, rval: &mut Box<Rvalue<'tcx>>, value: Const<'tcx>) {
```

Comment:
I think it should also work to turn the `rval` argument's type into `&mut Rvalue<'tcx>` in order to save the additional heap allocation (and the dropping of the old one).